src/test/groovy/graphql/schema/GraphQLSchemaTest.groovy

```diff
@@ -214,6 +214,46 @@ class GraphQLSchemaTest extends Specification {
         dogType.getName() == "Dog"
     }
 
+    def "issue with type references when original type is transformed away"() {
+        def sdl = '''
+            type Query {
+              # The b fields leads to the addition of the B type (actual definition)
+              b: B
+              # When we filter out the `b` field, we can still access B through A
+              # however they are GraphQLTypeReferences and not an actual GraphQL Object
+              a: A
+            } 
+
+            type A {
+              b: B
+            }
+
+            type B {
+              a: A
+              b: B
+            }
+        '''
+
+        when:
+        def schema = TestUtil.schema(sdl)
+        // When field `a` is filtered out
+        List<GraphQLFieldDefinition> fields = schema.queryType.fieldDefinitions.stream().filter({
+            it.name == "a"
+        }).collect(Collectors.toList())
+        // And we transform the schema's query root, the schema building
+        // will throw because type B won't be in the type map anymore, since
+        // there are no more actual B object types in the schema tree.
+        def transformed = schema.transform({
+            it.query(schema.queryType.transform({
+                it.replaceFields(fields)
+            }))
+        })
+
+        then:
+        transformed.containsType("B")
```

Comment:
I would expect this to pass since the schema pointers are `Query.a` -> `type A` -> `A.b` -> `type B`.  So this really feels like a bug.    Thanks for reporting

We will need to do some more investigation on why

Reply:
I started to look at this and I can indeed see it blowing up with something to do with `GraphQLTypeReference`s which is interesting

However it might be that this form of "schema editing" is not the right way.  I need to look further.

`GraphQLSchema newSchema = SchemaTransformer.transformSchema(schema, visitor);` is the best way to change a schema - it handles all of the invariants etc... versus directly editing the runtime graphql types.

For example it will fix up references and so on correctly.

As part of debugging this, I will write another test that shows how to do what you want (in filtering certain fields)

See https://www.graphql-java.com/documentation/schema and the section `Changing Schema`

Reply:
And this also fails - wow this is a real bug

See https://github.com/graphql-java/graphql-java/pull/2906 which is this PR plus another test